### PR TITLE
Add support for hardware copy protection dongles

### PR DESCRIPTION
**Pull request by medasaro**
This change adds support for hardware copy protection dongles.

Changes:
- Add a pulldown menu to the 'Advanced' tab that lets the user select a dongle in use
- Add a new file "CopyProgtectionDongles.cpp" that is a place to put drivers for these.
- Add a driver for the one known dongle we have now - Speed Star
- Modify Memory.cpp to alert the copy protection drivers when AN0-AN3 change
- Modify Joystick.cpp to allow PB0-PB2 to be "pushed" by the protection dongle.
- Misc.

Note: This is my first time contributing to this project, and my first time attempting to use git. Please go easy on me if I have made some dumb mistakes.

### CODE REVIEW

**Comment by medasaro** [comment]
Here is a simple test program I wrote to verify the functionality of the Speed Star protection dongle. Also attached is the expected output on real hardware.

-Matthew D'Asaro

[SoftKey Truth Table.zip](https://github.com/AppleWin/AppleWin/files/10231507/SoftKey.Truth.Table.zip)

**Comment by tomcw** [comment]
This is looking pretty good. Some minor things:

---

Rename these vars/funcs/enums/macros to use _non-truncated_ words:
- REGVALUE_CPYPRTDONGLE_TYPE -> REGVALUE_COPYPROTECTIONDONGLE_TYPE
- CPageAdvanced::InitCpyPrtDongleDropdownMenu() -> CPageAdvanced::InitCopyProtectionDongleDropdownMenu()
- m_CopyPrtDongleChoices[] -> m_CopyProtectionDongleChoices[]
- cpyPrtDongleType -> g_copyProtectionDongleType
- CPYPRTDONGLETYPE -> COPYPROTECTIONDONGLETYPE
- SetCpyPrtDongleType() -> SetCopyProtectionDongleType()
- GetCpyPrtDongleType() -> GetCopyProtectionDongleType()
- CopyProtDongleControl() -> CopyProtectionDongleControl()
- CopyProtDonglePB*() -> CopyProtectionDonglePB*()
- ...there may be others

---

Change all 3x CopyProtDonglePB*() funcs to return bool instead of DWORD.
Reason: SdsSpeedStar() returns a bool, but CopyProtDonglePB2() returns a DWORD here:
```
DWORD CopyProtDonglePB2(void)
{
	switch (cpyPrtDongleType)
	{
	case SDSSPEEDSTAR:	// Southwestern Data Systems SoftKey for Speed Star Applesoft Compiler
		return SdsSpeedStar();
```

---

PageAdvanced.cpp needs a final newline after that final `}` at the very end of the file.

---

Remove this code from Memory.cpp:

```
if (address >= 0xC058 && address <= 0xC05F)
		CopyProtDongleControl(address);	// Update AN0 - AN3 state for copy protection dongles
```

Instead call `MemGetAnnunciator(UINT annunciator)` to get the values of annunciators.

Then you can remove the 4x `static bool g_AN*` vars and `CopyProtDongleControl(const UINT uControl)` in "CopyProtectionDongles.cpp".

---

You've added a new .cpp file, so you'll need to include the AppleWin.vcxproj & AppleWin.vcxproj.filters files in this PR too.

---

btw, static bool g_AN* is internal state, which needs to be persisted or init'ed when loading a save-state. Probably the latter, since it's just a copy of Memory.cpp's `g_Annunciator[4]` variables. But if you can remove these 4x static bool g_AN*, as I suggest above, then even better.

But don't worry about save-state... I can add the support for this (and we'll also need to persist the `cpyPrtDongleType`).

Thanks!

**Comment by medasaro** [comment]
Thank you so much @tomcw for your helpful and detailed suggestions for improving my PR. I have implemented all of them except for "Change all 3x CopyProtDonglePB*() funcs to return bool instead of DWORD." If you note, I am using a return value of -1 to indicate that the copy protection dongle is not using a specific PB*. A return value of -1 is caught in Joystick.cpp: `				if(CopyProtectionDonglePB0() >= 0)				//If a copy protection dongle needs PB0, this overrides the joystick
					pressed = CopyProtectionDonglePB0();`

I am going to let you handle the save state as you offered - I looked at that code and it is not quite clear to me how it works.

Thanks,
Matthew D'Asaro

File: AppleWinExpress2019.vcxproj [hunk 1]
```diff
@@ -374,13 +376,13 @@
     <Keyword>Win32Proj</Keyword>
     <RootNamespace>AppleWin</RootNamespace>
     <ProjectName>AppleWin</ProjectName>
-    <WindowsTargetPlatformVersion>10.0</WindowsTargetPlatformVersion>
+    <WindowsTargetPlatformVersion>10.0.18362.0</WindowsTargetPlatformVersion>
```
Comment (tomcw) [inline AppleWinExpress2019.vcxproj:379]: Revert this.

File: AppleWinExpress2019.vcxproj [hunk 2]
```diff
@@ -374,13 +376,13 @@
...
     <ProjectName>AppleWin</ProjectName>
-    <WindowsTargetPlatformVersion>10.0</WindowsTargetPlatformVersion>
+    <WindowsTargetPlatformVersion>10.0.18362.0</WindowsTargetPlatformVersion>
   </PropertyGroup>
   <Import Project="$(VCTargetsPath)\Microsoft.Cpp.Default.props" />
   <PropertyGroup Condition="'$(Configuration)|$(Platform)'=='Debug|Win32'" Label="Configuration">
     <ConfigurationType>Application</ConfigurationType>
     <UseDebugLibraries>true</UseDebugLibraries>
-    <PlatformToolset>v142</PlatformToolset>
+    <PlatformToolset>v141</PlatformToolset>
```
Comment (tomcw) [inline AppleWinExpress2019.vcxproj:385]: Revert this.

File: libyaml/win32/yaml2019.vcxproj [hunk 3]
```diff
@@ -42,13 +42,13 @@
     <Keyword>Win32Proj</Keyword>
     <RootNamespace>yaml2013</RootNamespace>
     <ProjectName>yaml</ProjectName>
-    <WindowsTargetPlatformVersion>10.0</WindowsTargetPlatformVersion>
+    <WindowsTargetPlatformVersion>10.0.18362.0</WindowsTargetPlatformVersion>
```
Comment (tomcw) [inline libyaml/win32/yaml2019.vcxproj:45]: Revert this.

File: libyaml/win32/yaml2019.vcxproj [hunk 4]
```diff
@@ -42,13 +42,13 @@
...
     <ProjectName>yaml</ProjectName>
-    <WindowsTargetPlatformVersion>10.0</WindowsTargetPlatformVersion>
+    <WindowsTargetPlatformVersion>10.0.18362.0</WindowsTargetPlatformVersion>
   </PropertyGroup>
   <Import Project="$(VCTargetsPath)\Microsoft.Cpp.Default.props" />
   <PropertyGroup Condition="'$(Configuration)|$(Platform)'=='Debug|Win32'" Label="Configuration">
     <ConfigurationType>StaticLibrary</ConfigurationType>
     <UseDebugLibraries>true</UseDebugLibraries>
-    <PlatformToolset>v142</PlatformToolset>
+    <PlatformToolset>v141</PlatformToolset>
```
Comment (tomcw) [inline libyaml/win32/yaml2019.vcxproj:51]: Revert this.

File: test/TestCPU6502/TestCPU6502-vs2019.vcxproj [hunk 5]
```diff
@@ -31,13 +31,13 @@
     <Keyword>Win32Proj</Keyword>
     <RootNamespace>TestCPU6502vs2013</RootNamespace>
     <ProjectName>TestCPU6502</ProjectName>
-    <WindowsTargetPlatformVersion>10.0</WindowsTargetPlatformVersion>
+    <WindowsTargetPlatformVersion>10.0.18362.0</WindowsTargetPlatformVersion>
```
Comment (tomcw) [inline test/TestCPU6502/TestCPU6502-vs2019.vcxproj:34]: Revert this.

File: test/TestCPU6502/TestCPU6502-vs2019.vcxproj [hunk 6]
```diff
@@ -31,13 +31,13 @@
...
     <ProjectName>TestCPU6502</ProjectName>
-    <WindowsTargetPlatformVersion>10.0</WindowsTargetPlatformVersion>
+    <WindowsTargetPlatformVersion>10.0.18362.0</WindowsTargetPlatformVersion>
   </PropertyGroup>
   <Import Project="$(VCTargetsPath)\Microsoft.Cpp.Default.props" />
   <PropertyGroup Condition="'$(Configuration)|$(Platform)'=='Debug|Win32'" Label="Configuration">
     <ConfigurationType>Application</ConfigurationType>
     <UseDebugLibraries>true</UseDebugLibraries>
-    <PlatformToolset>v142</PlatformToolset>
+    <PlatformToolset>v141</PlatformToolset>
```
Comment (tomcw) [inline test/TestCPU6502/TestCPU6502-vs2019.vcxproj:40]: Revert this.

File: zip_lib/zip_lib2019.vcxproj [hunk 7]
```diff
@@ -38,13 +38,13 @@
     <Keyword>Win32Proj</Keyword>
     <RootNamespace>zip_lib</RootNamespace>
     <ProjectName>zip_lib</ProjectName>
-    <WindowsTargetPlatformVersion>10.0</WindowsTargetPlatformVersion>
+    <WindowsTargetPlatformVersion>10.0.18362.0</WindowsTargetPlatformVersion>
```
Comment (tomcw) [inline zip_lib/zip_lib2019.vcxproj:41]: Revert this.

File: zip_lib/zip_lib2019.vcxproj [hunk 8]
```diff
@@ -38,13 +38,13 @@
...
     <ProjectName>zip_lib</ProjectName>
-    <WindowsTargetPlatformVersion>10.0</WindowsTargetPlatformVersion>
+    <WindowsTargetPlatformVersion>10.0.18362.0</WindowsTargetPlatformVersion>
   </PropertyGroup>
   <Import Project="$(VCTargetsPath)\Microsoft.Cpp.Default.props" />
   <PropertyGroup Condition="'$(Configuration)|$(Platform)'=='Debug|Win32'" Label="Configuration">
     <ConfigurationType>StaticLibrary</ConfigurationType>
     <UseDebugLibraries>true</UseDebugLibraries>
-    <PlatformToolset>v142</PlatformToolset>
+    <PlatformToolset>v141</PlatformToolset>
```
Comment (tomcw) [inline zip_lib/zip_lib2019.vcxproj:47]: Revert this.

File: zlib/zlib-Express2019.vcxproj [hunk 9]
```diff
@@ -53,13 +53,13 @@
     <Keyword>Win32Proj</Keyword>
     <RootNamespace>zlib</RootNamespace>
     <ProjectName>zlib</ProjectName>
-    <WindowsTargetPlatformVersion>10.0</WindowsTargetPlatformVersion>
+    <WindowsTargetPlatformVersion>10.0.18362.0</WindowsTargetPlatformVersion>
```
Comment (tomcw) [inline zlib/zlib-Express2019.vcxproj:56]: Revert this.

File: zlib/zlib-Express2019.vcxproj [hunk 10]
```diff
@@ -53,13 +53,13 @@
...
     <ProjectName>zlib</ProjectName>
-    <WindowsTargetPlatformVersion>10.0</WindowsTargetPlatformVersion>
+    <WindowsTargetPlatformVersion>10.0.18362.0</WindowsTargetPlatformVersion>
   </PropertyGroup>
   <Import Project="$(VCTargetsPath)\Microsoft.Cpp.Default.props" />
   <PropertyGroup Condition="'$(Configuration)|$(Platform)'=='Debug|Win32'" Label="Configuration">
     <ConfigurationType>StaticLibrary</ConfigurationType>
     <UseDebugLibraries>true</UseDebugLibraries>
-    <PlatformToolset>v142</PlatformToolset>
+    <PlatformToolset>v141</PlatformToolset>
```
Comment (tomcw) [inline zlib/zlib-Express2019.vcxproj:62]: Revert this.

**Review by tomcw — CHANGES_REQUESTED** [review]
With the addition of those .vcxproj files, Visual Studio has "helpfully" made some changes that I'll need you to undo:
(eg. for `WindowsTargetPlatformVersion` and `PlatformToolset`)

- AppleWinExpress2019.vcxproj
- yaml2019.vcxproj
- TestCPU6502-vs2019.vcxproj
- zip_lib2019.vcxproj
- zlib-Express2019.vcxproj

I've annotated with "Revert this." comments.

**Comment by medasaro** [comment]
I believe I have resolved the corrupted project file issue with my latest commit. Sorry about that.

-Matthew D'Asaro

**Comment by medasaro** [comment]
Hmm. I don't think I meant to "close" this. Reopening.

-Matthew D'Asaro

**Review by tomcw — APPROVED** [review]

**Comment by tomcw** [comment]
Thanks @medasaro - successfully merged into the main line!
btw, I'll add a credit to you in the docs & history for this contribution.

**Pull request merged**